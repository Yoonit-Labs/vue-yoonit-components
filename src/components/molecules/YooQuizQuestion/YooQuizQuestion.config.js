module.exports = {
  textPosition: {
    options: [
      'left',
      'right'
    ]
  }
}
