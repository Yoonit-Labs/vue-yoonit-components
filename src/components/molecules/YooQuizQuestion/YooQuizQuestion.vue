<template lang="pug">
  YooGridLayout(
    rows="*"
    :cols="doSetRows(takeAnswers)"
    :class="['yoo-quiz']"
  )
    p(
      row="1"
      col="1"
      :class="['yoo-quiz__question']"
    ) {{ takeQuestion.text.ptBR }}
      span(
        v-if="takeQuestion.required"
      ) *

    YooSwitchCard(
      v-for="(ans, index) in takeQuestion.answers"
      :key="'radio-' + takeQuestion.id + '-' + ans.id + '-' + index"
      :row="index+2"
      :col="1"
      :class="['yoo-quiz__switch-card']"
      :textPosition="textPosition"
      :text="ans.text.ptBR"
      :checked="doTakeStatus(index)"
      @response='doTap($event, takeQuestion.id, index, ans.override, takeQuestion.type)'
    )

</template>

<script>

import YooGridLayout from '@/components/quarks/YooGridLayout/YooGridLayout.vue'
import YooSwitchCard from '@/components/molecules/YooSwitchCard/YooSwitchCard.vue'
import PropsConfig from '@/components/molecules/YooQuizQuestion/YooQuizQuestion.config'

export default {
  name: 'YooQuizQuestion',
  props: {
    textPosition: {
      type: String,
      default: 'right',
      validator: value => PropsConfig.textPosition.options.includes(value)
    },
    question: {
      type: Object,
      required: true
    },
    answers: {
      type: Array,
      required: true
    }
  },
  components: {
    YooGridLayout,
    YooSwitchCard
  },
  data: function () {
    return {
      localAnswers: []
    }
  },
  beforeCreate () {},
  created () {},
  beforeMount () {},
  mounted () {
    this.localAnswers = this.answers
  },
  beforeUpdate () {},
  updated () {},
  beforeDestroy () {},
  destroyed () {},
  computed: {
    takeQuestion () {
      return this.question
    },
    takeAnswers () {
      return this.question.answers || []
    }
  },
  methods: {
    doSetRows (answers) {
      return answers
        .map(elm => 'auto')
        .join(',')
    },
    doTap (event, questionId, answerId, override, type) {
      if ((override && event) || type === 'radio') {
        for (var answer in this.localAnswers) {
          this.$set(this.localAnswers[answer], 'status', false)
        }
      } else {
        for (var otherAnswer in this.localAnswers) {
          if (this.localAnswers[otherAnswer].override === true) {
            this.localAnswers[otherAnswer].status = false
          }
        }
      }

      this.$set(this.localAnswers[answerId], 'status', event)
      return this.$emit('tapChoice', this.localAnswers)
    },
    doTakeStatus (ans) {
      return this.localAnswers && this.localAnswers[ans]
        ? this.localAnswers[ans].status
        : false
    }
  },
  filters: {},
  watch: {}
}
</script>

<style src="./YooQuizQuestion.sass" lang="sass" scoped></style>
