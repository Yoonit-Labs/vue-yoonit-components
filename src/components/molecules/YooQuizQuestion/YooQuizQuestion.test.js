import { mount } from '@vue/test-utils'
import PropsConfig from './YooQuizQuestion.config'
import YooQuizQuestion from './YooQuizQuestion.vue'

const SlotText = 'Default Slot Text'
const answers = [
  {
    id: 'ans1',
    text: {
      ptBR: 'Cansaço'
    },
    weight: 3
  },
  {
    id: 'ans2',
    text: {
      ptBR: 'Corrimento Nasal (Coriza)'
    },
    weight: 3,
    status: true
  }
]

const question = {
  id: 'covid-cl1',
  text: {
    ptBR: 'Marque os sintomas sentidos nas últimas horas'
  },
  visible: true,
  type: 'check',
  answers
}

const mountComponent = () => {
  return mount(YooQuizQuestion, {
    slots: { default: SlotText },
    propsData: {
      question,
      answers
    }
  })
}

describe('YooQuizQuestion Component', () => {
  let wrapper
  beforeEach(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(0.123456789)
    wrapper = mountComponent()
  })

  afterEach(() => {
    jest.spyOn(global.Math, 'random').mockRestore()
  })

  it('Matches Snapshot', async () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Loads the Component HTML', () => {
    expect(wrapper.classes('yoo-quiz')).toBe(true)
    expect(wrapper.find('.yoo-quiz__question').exists()).toBe(true)
    expect(wrapper.find('.yoo-quiz__switch-card').exists()).toBe(true)
  })

  describe('Props', () => {
    describe('textPosition', () => {
      it('Has a valid default value', () => {
        expect(PropsConfig.textPosition.options.includes(YooQuizQuestion.props.textPosition.default)).toBe(true)
      })
      PropsConfig.textPosition.options.forEach(textPosition => {
        it('Loads the Component HTML', async () => {
          await wrapper.setProps({ textPosition })
          expect(wrapper.find(`.yoo-check--text-${textPosition}`).exists()).toBe(true)
        })
      })
    })

    describe('question required', () => {
      it('Does not `*` if question required', async () => {
        expect(wrapper.text()).not.toMatch('*')
      })
      it('Show `*` if question required', async () => {
        await wrapper.setProps({ question: { ...question, required: true } })
        expect(wrapper.text()).toMatch('*')
      })
    })

    describe('answers', () => {
      it('Includes class: .yoo-quiz__switch-card when array answers has lenght', async () => {
        expect(wrapper.find('.yoo-quiz__switch-card').exists()).toBe(true)
      })
      it('Does not include class: .yoo-quiz__switch-card when array answers does not has lenght', async () => {
        await wrapper.setProps({
          answers: [],
          question: {
            id: 'covid-cl1',
            text: {
              ptBR: 'Marque os sintomas sentidos nas últimas horas'
            },
            visible: true,
            type: 'check',
            answers: []
          }
        })
        expect(wrapper.find('.yoo-quiz__switch-card').exists()).toBe(false)
      })
    })
  }) // describe Props

  describe('Events', () => {
    describe('Click', () => {
      it('Emits Click Event', async () => {
        await wrapper.find('.yoo-quiz__switch-card').vm.$emit('response')
        expect(wrapper.emitted()).toHaveProperty('tapChoice')
      })
    })
  }) // describe Events
})
