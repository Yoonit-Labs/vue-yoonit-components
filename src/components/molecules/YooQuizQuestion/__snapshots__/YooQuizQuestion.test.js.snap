// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooQuizQuestion Component Matches Snapshot 1`] = `
<div class="yoo-grid-layout yoo-grid-layout--h-stretch yoo-grid-layout--v-stretch yoo-quiz" style="grid-template-rows: auto; grid-template-columns: auto auto;">
  <p row="1" col="1" class="yoo-quiz__question" style="grid-row: 1; grid-column: 1;">Marque os sintomas sentidos nas últimas horas
    <!---->
  </p>
  <div class="yoo-flex-layout m__t--s m__b--s m__r--l m__l--l p__t--l p__r--l p__b--l p__l--l yoo-switch-card yoo-quiz__switch-card" style="flex-direction: row; flex-wrap: nowrap; justify-content: center; align-items: stretch; align-content: stretch; gap: 16px; grid-row: 2; grid-column: 1;" row="2" col="1">
    <div class="yoo-checkbox yoo-checkbox--switch-card yoo-switch-card__checkbox"><input type="checkbox" id="ycid-0889654"><label class="yoo-check fas fa-check yoo-check--medium yoo-check--checked yoo-check--switch-card yoo-check--text-right" for="ycid-0889654">
        <p>Cansaço</p>
      </label></div>
  </div>
  <div class="yoo-flex-layout m__t--s m__b--s m__r--l m__l--l p__t--l p__r--l p__b--l p__l--l yoo-switch-card yoo-switch-card--active yoo-quiz__switch-card" style="flex-direction: row; flex-wrap: nowrap; justify-content: center; align-items: stretch; align-content: stretch; gap: 16px; grid-row: 3; grid-column: 1;" row="3" col="1">
    <div class="yoo-checkbox yoo-checkbox--switch-card yoo-switch-card__checkbox"><input type="checkbox" id="ycid-6020099"><label class="yoo-check fas fa-check yoo-check--medium yoo-check--checked yoo-check--switch-card yoo-check--text-right" for="ycid-6020099">
        <p>Corrimento Nasal (Coriza)</p>
      </label></div>
  </div>
</div>
`;
